File: C/2303-calculate-amount-paid-in-taxes/main.go
package main

func calculateTax(brackets [][]int, income int) float64 {
	prev := 0

	var tax float64 = 0.0
	for i := 0; i < len(brackets); i++ {
		if brackets[i][0] >= income {
			break
		}
		remain := brackets[i][0] - prev
		if remain > income-prev {
			remain = income - prev
		}
		prev = brackets[i][0]
		tax += float64(remain*brackets[i][1]) / 100
	}
	return tax
}

File: C/2303-calculate-amount-paid-in-taxes/main.rs
struct Solution {}

impl Solution {
  pub fn calculate_tax(brackets: Vec<Vec<i32>>, income: i32) -> f64 {
    let mut agg: i32 = 0;
    let mut sum: f64 = 0.0;
    let mut idx: usize = 0;
    while agg <= income && idx < brackets.len() {
      let mut v = brackets[idx][0] - agg;
      if v > income - agg {
        v = income - agg;
      }

      sum += v as f64 * brackets[idx][1] as f64 / 100 as f64;
      agg = brackets[idx][0];
      idx += 1;
    }

    sum
  }
}
